src/v/transform/rpc/client.cc

```diff
@@ -0,0 +1,154 @@
+#include <seastar/core/chunked_fifo.hh>
+#include <seastar/core/future.hh>
+#include <seastar/core/loop.hh>
+#include <seastar/core/semaphore.hh>
+#include <seastar/core/smp.hh>
+#include <seastar/core/when_all.hh>
+#include <seastar/util/noncopyable_function.hh>
+
+#include <absl/container/btree_map.h>
+#include <absl/container/flat_hash_map.h>
+#include <boost/fusion/sequence/intrinsic/back.hpp>
+
+#include <algorithm>
+#include <chrono>
+#include <exception>
+#include <iterator>
+#include <stdexcept>
+#include <system_error>
+#include <utility>
+#include <vector>
+
+namespace transform::rpc {
+
+namespace {
+constexpr auto timeout = std::chrono::seconds(1);
+
+cluster::errc map_errc(std::error_code ec) {
+    if (ec.category() == cluster::error_category()) {
+        return static_cast<cluster::errc>(ec.value());
+    } else if (ec.category() == raft::error_category()) {
+        auto raft_ec = static_cast<raft::errc>(ec.value());
+        switch (raft_ec) {
+        case raft::errc::not_leader:
+        case raft::errc::leadership_transfer_in_progress:
+            return cluster::errc::not_leader;
+        default:
+            vlog(log.error, "unexpected transform produce error: {}", raft_ec);
+            break;
+        }
+    } else if (ec.category() == ::rpc::error_category()) {
+        auto rpc_ec = static_cast<::rpc::errc>(ec.value());
+        switch (rpc_ec) {
+        case ::rpc::errc::client_request_timeout:
+        case ::rpc::errc::connection_timeout:
+        case ::rpc::errc::disconnected_endpoint:
+            return cluster::errc::timeout;
+        default:
+            vlog(log.error, "unexpected transform produce error: {}", rpc_ec);
+            break;
+        }
+    } else {
+        vlog(log.error, "unexpected transform produce error: {}", ec);
+    }
+    return cluster::errc::timeout;
+}
+} // namespace
+
+client::client(
+  model::node_id self,
+  ss::sharded<cluster::partition_leaders_table>* l,
+  ss::sharded<::rpc::connection_cache>* c,
+  ss::sharded<local_service>* s)
+  : _self(self)
+  , _leaders(l)
+  , _connections(c)
+  , _local_service(s) {}
+
+ss::future<cluster::errc> client::produce(
+  model::topic_partition tp, ss::chunked_fifo<model::record_batch> batches) {
+    vlog(log.trace, "producing {} batches to {}", batches.size(), tp);
+    auto leader = _leaders->local().get_leader(
+      model::topic_namespace_view(model::kafka_namespace, tp.topic),
+      tp.partition);
+    if (!leader) {
+        co_return cluster::errc::not_leader;
+    }
+    produce_request req;
+    req.topic_data.emplace_back(std::move(tp), std::move(batches));
+    req.timeout = timeout;
+    auto reply = co_await (
+      *leader == _self ? do_local_produce(std::move(req))
+                       : do_remote_produce(*leader, std::move(req)));
+    vassert(
+      reply.results.size() == 1,
+      "expected a single result: {}",
+      reply.results.size());
+
+    co_return reply.results.front().err;
+}
```

Comment:
question: would it be worthwhile to attempt a retry loop in here in case partition leadership changes while message is in flight?

Reply:
Really comes down to how similar to a kafka client we want this to be

Reply:
Right now I'm aiming for as simple as possible and I'll let the upstream processing and error handling retry, but I certainly will be adding some retry/batching logic in the client.

Reply:
So yes it's worthwhile, I'll probably put up another PR for that later.

Reply:
> So yes it's worthwhile, I'll probably put up another PR for that later.

Or wait for the next service to do it when they need it (*cough*)